packages/web3-providers-ipc/src/index.ts

```diff
@@ -0,0 +1,193 @@
+			throw new InvalidClientError(this._socketPath);
+		}
+
+		try {
+			this._addSocketListeners();
+			this._socket.connect({ path: this._socketPath });
+		} catch (e) {
+			throw new InvalidConnectionError(this._socketPath);
+		}
+	}
+
+	public disconnect(): void {
+		this._requestQueue.clear();
+		this._removeSocketListeners();
+		this._socket.end();
+	}
+
+	public reset(): void {
+		this._requestQueue.clear();
+
+		this._removeSocketListeners();
+		this._addSocketListeners();
+	}
+
+	// eslint-disable-next-line @typescript-eslint/require-await
+	public async request<
+		Method extends Web3APIMethod<API>,
+		ResponseType = Web3APIReturnType<API, Method>,
+	>(request: Web3APIPayload<API, Method>): Promise<JsonRpcResponse<ResponseType>> {
+		if (this._socket === undefined) throw new Error('IPC connection is undefined');
+
+		if (request.id === undefined) throw new Error('Request Id not defined');
+
+		if (this.getStatus() !== 'connected') {
+			throw new ConnectionNotOpenError();
+		}
+
+		try {
+			this._socket.write(JSON.stringify(request));
+			const defPromise = new DeferredPromise<JsonRpcResponse<ResponseType>>();
+			this._requestQueue.set(request.id, defPromise);
+			return defPromise;
+		} catch (error) {
+			this._requestQueue.delete(request.id);
+			throw error;
+		}
+	}
+
+	public removeAllListeners(type: string): void {
+		this._emitter.removeAllListeners(type);
+	}
+
+	private _onMessage(e: Buffer | string): void {
+		const result = typeof e === 'string' ? e : e.toString('utf8');
+
+		const response = JSON.parse(result) as
+			| JsonRpcResponseWithError
+			| JsonRpcResponseWithResult
+			| JsonRpcNotification;
+
+		if ('method' in response && response.method.endsWith('_subscription')) {
+			this._emitter.emit('message', null, response);
+			return;
+		}
+
+		if (response.id && this._requestQueue.has(response.id)) {
+			const requestItem = this._requestQueue.get(response.id);
+
+			if ('result' in response && response.result !== undefined) {
+				this._emitter.emit('message', null, response);
+				requestItem?.resolve(response);
+			} else if ('error' in response && response.error !== undefined) {
+				this._emitter.emit('message', response, null);
+				requestItem?.reject(response);
+			}
+
+			this._requestQueue.delete(response.id);
+		}
+	}
+
+	private _clearQueues(event?: CloseEvent) {
+		if (this._requestQueue.size > 0) {
+			this._requestQueue.forEach((request: DeferredPromise<unknown>, key: JsonRpcId) => {
+				request.reject(new ConnectionNotOpenError(event));
+				this._requestQueue.delete(key);
+			});
+		}
+
+		this._removeSocketListeners();
+	}
+
+	private _onConnect() {
+		this._connectionStatus = 'connected';
+		this._emitter.emit('connect');
+	}
+
+	private _onDisconnect(): void {
+		this._connectionStatus = 'disconnected';
+		this._emitter.emit('disconnect');
+	}
+
+	private _onClose(event: CloseEvent): void {
+		this._clearQueues(event);
+		this._removeSocketListeners();
+	}
+
+	private _addSocketListeners(): void {
```

Comment:
[1.x](https://github.com/ChainSafe/web3.js/blob/1.x/packages/web3-providers-ipc/src/index.js#L86) had listeners `error` and `timeout`, should we add them?

Reply:
As per the docs the `timeout` is really just an notification of inactivity of idle socket. So we should not be worried about it. 

> Emitted if the socket times out from inactivity. This is only to notify that the socket has been idle. The user must manually close the connection.


The `error` event is unnecessary in case we have `close` event, as `close` will be triggered directly after. 

> Emitted when an error occurs. The 'close' event will be called directly following this event.